### ISSUE TITLE
Caught exceptions still take down server

### ISSUE THREAD

**Issue by codinronan** [stripe/stripe-node#594]
Hi there,

So I did search the repo for this issue but it seems it's only been reported once or twice, and either was marked fixed or the reporter never replied again.

I have what seems to be a basic issue with catching exceptions thrown by Stripe requests. First, the needful:

Node: 10.12
Stripe: 6.28.0

Ok so first of all this is the offending piece of code.

We had a configuration error this morning and our website got deployed to production for a few minutes with a test publishable key. During that time, one transaction was received, and as  a result of the key mismatch (server has a production key, website creates sources with a test key), the following operation threw an error:

```
async function addCustomerSource(stripeCustomerId, source) {
  try {
    const customer = await stripe.customers.update(stripeCustomerId, { source });
    return { customer };
  } catch (error) {
    logger.error('Error adding source to customer:', null, error, { stripeCustomerId });
    return { error };
  }
}
```

As you can see I am catching the exception and reporting it. Now, it *is* actually caught. My log then spits out the following information, partially redacted:

```
2019-04-19T20:14:35.658Z - ERROR: - Error adding source to customer: - [[log data redacted]] 
{ Error: No such source: src_1E****redacted******Um; 
a similar object exists in test mode, but a live mode key was used to make this request.
    at Constructor._Error (/home/nowuser/src/node_modules/stripe/lib/Error.js:12:17)
    at Constructor (/home/nowuser/src/node_modules/stripe/lib/utils.js:138:13)
    at new Constructor (/home/nowuser/src/node_modules/stripe/lib/utils.js:138:13)
    at Function.StripeError.generate (/home/nowuser/src/node_modules/stripe/lib/Error.js:57:12)
    at IncomingMessage.<anonymous> (/home/nowuser/src/node_modules/stripe/lib/StripeResource.js:163:39)
    at IncomingMessage.emit (events.js:187:15)
    at IncomingMessage.EventEmitter.emit (domain.js:460:23)
    at endReadableNT (_stream_readable.js:1081:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  type: 'StripeInvalidRequestError',
  stack:
   'Error: No such source: src_1ER36TIlsyx32qczFzHqFIUm; a similar object exists in test mode, but a live mode key was used to make this request.\n
    at Constructor._Error (/home/nowuser/src/node_modules/stripe/lib/Error.js:12:17)\n
    at Constructor (/home/nowuser/src/node_modules/stripe/lib/utils.js:138:13)\n
   at new Constructor (/home/nowuser/src/node_modules/stripe/lib/utils.js:138:13)\n
    at Function.StripeError.generate (/home/nowuser/src/node_modules/stripe/lib/Error.js:57:12)\n
    at IncomingMessage.<anonymous> (/home/nowuser/src/node_modules/stripe/lib/StripeResource.js:163:39)\n
    at IncomingMessage.emit (events.js:187:15)\n
    at IncomingMessage.EventEmitter.emit (domain.js:460:23)\n
    at endReadableNT (_stream_readable.js:1081:12)\n
    at process._tickCallback (internal/process/next_tick.js:63:19)',
  rawType: 'invalid_request_error',
  code: 'resource_missing',
  param: 'source',
  message:
   'No such source: src_1E****redacted******Um; a similar object exists in test mode, but a live mode key was used to make this request.',
  detail: undefined,
  raw:
   { code: 'resource_missing',
     doc_url: 'https://stripe.com/docs/error-codes/resource-missing',
     message:
      'No such source: src_1E****redacted******Um; a similar object exists in test mode, but a live mode key was used to make this request.',
     param: 'source',
     type: 'invalid_request_error',
     headers:
      { server: 'nginx',
        date: 'Fri, 19 Apr 2019 20:14:35 GMT',
        'content-type': 'application/json',
        'content-length': '336',
        connection: 'keep-alive',
        'access-control-allow-credentials': 'true',
        'access-control-allow-methods': 'GET, POST, HEAD, OPTIONS, DELETE',
        'access-control-allow-origin': '*',
        'access-control-expose-headers':
         'Request-Id, Stripe-Manage-Version, X-Stripe-External-Auth-Required, X-Stripe-Privileged-Session-Required',
        'access-control-max-age': '300',
        'cache-control': 'no-cache, no-store',
        'request-id': 'req_hQPNBi8vGV6LZL',
        'stripe-version': '2019-03-14',
        'strict-transport-security': 'max-age=31556926; includeSubDomains; preload' },
     statusCode: 400,
     requestId: 'req_hQPNBi8vGV6LZL' },
  headers:
   { server: 'nginx',
     date: 'Fri, 19 Apr 2019 20:14:35 GMT',
     'content-type': 'application/json',
     'content-length': '336',
     connection: 'keep-alive',
     'access-control-allow-credentials': 'true',
     'access-control-allow-methods': 'GET, POST, HEAD, OPTIONS, DELETE',
     'access-control-allow-origin': '*',
     'access-control-expose-headers':
      'Request-Id, Stripe-Manage-Version, X-Stripe-External-Auth-Required, X-Stripe-Privileged-Session-Required',
     'access-control-max-age': '300',
     'cache-control': 'no-cache, no-store',
     'request-id': 'req_hQPNBi8vGV6LZL',
     'stripe-version': '2019-03-14',
     'strict-transport-security': 'max-age=31556926; includeSubDomains; preload' },
  requestId: 'req_hQPNBi8vGV6LZL',
  statusCode: 400 }
Killed
npm ERR! code ELIFECYCLE
npm ERR! errno 137
npm ERR! server@1.16.20 start: `NODE_ENV=production node --max-old-space-size=2048 server.js`
npm ERR! Exit status 137
npm ERR! 
npm ERR! Failed at the server@1.16.20 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/nowuser/.npm/_logs/2019-04-19T20_15_31_571Z-debug.log
exit code: 0
```

However as you note, this exception actually took down the server - even though it was caught.

We then reproduced this locally, *very* reliably. A) Adding a source, or B) attempting to charge a source, in each case, if the transaction fails, the error is thrown, caught, and reported, and then the server immediately crashes.

There are no additional errors. We have full error and kill signal handling, and none of those handlers fire: that means this is not only an unhandled exception, it's a hard crash.

Possibly relevant:
This server uses Mongoose to communicate with a remote Mongo server (but therefore does have the driver running locally).

We have two other modules that pretty heavily change the operating environment:
1. Bugsnag (for capturing errors; the above was reported to bugsnag)
2. Split.io for feature switch management. Split.io I can see in the terminal is actually spawning other processes so it seems relevant here.

Given how reliable this crash is, It *seems* on the surface related to Stripe throwing exceptions. Perhaps I need to be using the `Promise(...).then(..).catch(...)` form of these functions rather than using the `async/await` syntactic sugar?

Edit:
This one seems related: https://github.com/stripe/stripe-node/issues/440

**Comment by ob-stripe** [comment]
Hi @codinronan. I don't think this is an issue with stripe-node itself. As you noticed, the exception is caught and handled.

"Exit status 137" indicates that the process was killed by the Linux kernel because the server was running out of memory. Maybe there's a memory leak in your code or one of your third-party modules, that only occur when processing an error?

**Comment by codinronan** [comment]
@ob-stripe Hey there thanks for replying. Yeah, it doesn't make any sense. JS doesn't change the rules on the fly.

I'm going to disable the other modules and report back - will let you know...

**Comment by ob-stripe** [comment]
I would also recommend monitoring memory usage in your tests, to confirm this is an OOM error.

**Comment by codinronan** [comment]
@ob-stripe Following up on this:

I check if OOM was the crash reason and it wasn't. 

What I did find was an error in my route handling that meant for certain kinds of errors (including this one) there was no error route defined, which meant that Express would use its default one. Since the error object thrown is not a normal error, it looks like there's a hard crash somewhere in Express itself (that's where I stopped debugging).

I fixed the route handler and can no longer repro the issue. Seems this can be closed.
